File: change.pl
#!/usr/bin/env perl

use strict;
use warnings;

my $gh_url = "https://GITHUB_REPO";
my $target_dir = "FOLDER_TO_KEEP_UPDATED";

my $temp_repo_location = "/tmp/temp_repo_location";

# clone gh repo to temp dir
system("rm", "-rf", "$temp_repo_location");
system("git", "clone", "$gh_url", "$temp_repo_location");

# don't compare .git files
system("rm", "-rf", "$temp_repo_location/.git");

# compare differences
system("diff", "-r", "$target_dir", "$temp_repo_location");

if ($?) {
  system("rm", "-rf", "$target_dir");
  system("cp", "-r", "$temp_repo_location", "$target_dir");
  print "Detected a difference!\n";
}

exit 0;

File: change.py
import os
import subprocess

any_updates = False

def process_repo(url, name):
    repo_location = "/tmp/" + name
    if (os.path.isdir(repo_location)):
        # if dir exists - check update
        check_for_updates(repo_location, name)
    else:
        # clone gh repo to temp dir
        subprocess.call(["/usr/bin/git", "clone", url, repo_location]);
        install(repo_location)

def check_for_updates(repo_path, name):
    completed_process = subprocess.call(["/usr/bin/git", "diff", "remotes/origin/HEAD"], cwd=repo_path);
    if (completed_process.returncode != 0 ):
        # Update
        completed_process = subprocess.call(["/usr/bin/git", "pull"], cwd=repo_path);
        install(repo_path, name)
        print("updating " + name)

def install(repo_path):
    subprocess.call([repo_path + "/install.sh"])
    any_updates = True

# read tracking file
track = open("/usr/local/etc/track.csv", "r")
for line in track:
    tracking_data = line.split(',')
    name = tracking_data[0]
    gh_url = tracking_data[1].strip()
    process_repo(gh_url, name)

track.close()

if (any_updates):
    subprocess.call(["/sbin/reboot"])
